### PR TITLE
SCons: Streamline Vulkan buildsystem + fixups

### PR DESCRIPTION
**Pull request by akien-mga**
- Renamed option to `builtin_vulkan`, since that's the name of the
  library and if we were to add new components, we'd likely use that
  same option.
- Merge `vulkan_loader/SCsub` in `vulkan/SCsub`.
- Accordingly, don't use built-in Vulkan headers when not building
  against the built-in loader library.
- Drop Vulkan registry which we don't appear to need currently.
- Style and permission fixes.

 @bruvzg Not sure about removing the registry, do we need it for something? We don't seem to use `vk.xml` directly, so I assume not, but maybe it was added to be used during development?

Please also test on MSVC and macOS, I only tested:
- Linux with `builtin_vulkan=yes` and `builtin_vulkan=no`.
- MinGW from Linux with `builtin_vulkan=yes`.

### CODE REVIEW

**Review by bruvzg — APPROVED** [review]
On macOS everything seems to be OK.

> @bruvzg Not sure about removing the registry, do we need it for something? We don't seem to use vk.xml directly, so I assume not, but maybe it was added to be used during development?

`vk.xml` and python scripts in the `registry` folder are used to generate some of Vulkan headers (all pre-generated headers are included in the `loader` folder).

**Comment by akien-mga** [comment]
> `vk.xml` and python scripts in the `registry` folder are used to generate some of Vulkan headers (all pre-generated headers are included in the `loader` folder).

If we follow https://github.com/KhronosGroup/Vulkan-Loader/blob/master/BUILD.md#building-dependent-repositories-with-known-good-revisions and https://github.com/KhronosGroup/Vulkan-Loader/blob/master/BUILD.md#generated-source-code when updating the loader, we should get the matching headers + generated sources, no?

**Comment by bruvzg** [comment]
> If we follow https://github.com/KhronosGroup/Vulkan-Loader/blob/master/BUILD.md#building-dependent-repositories-with-known-good-revisions and https://github.com/KhronosGroup/Vulkan-Loader/blob/master/BUILD.md#generated-source-code when updating the loader, we should get the matching headers + generated sources, no?

Yep that's exactly what's generated, but it was added after 1.1.114 (my PR use 1.1.113 release), previously there were no generated headers.

**Comment by akien-mga** [comment]
I pushed an update to `sdk-1.1.114.0`. I kept the generated files in the same `loader/` folder instead of upstream's `loader/generated/` file structure, mostly to save having to append yet another long include path to the main `env` :)

File: thirdparty/vulkan/loader/loader.c [hunk 1]
```diff
@@ -63,6 +69,11 @@
 #include <cfgmgr32.h>
 #include <initguid.h>
 #include <devpkey.h>
+#include <winternl.h>
+#include <d3dkmthk.h>
```
Comment (akien-mga) [inline thirdparty/vulkan/loader/loader.c]: This makes things tricky for cross-compilation from Linux as this header is not part of mingw but is provided by WINE. In my distro it's at `/usr/include/wine/windows/ddk/d3dkmthk.h` which is not in the default include paths, so I'll have to hardcode this extra include path.

Comment (akien-mga) [reply thirdparty/vulkan/loader/loader.c]: And that header requires `/usr/include/wine/windows/d3dukmdt.h` too, so another path to hardcode \o/

Comment (akien-mga) [reply thirdparty/vulkan/loader/loader.c]: Fun, on Debian-based distros the `wine` part of the path is duplicated: https://packages.ubuntu.com/disco/amd64/libwine-dev/filelist

And it actually doesn't build even then, I get a ton of build errors typical to windows headers being included in the wrong order...

**Comment by akien-mga** [comment]
I removed the version update for now as it doesn't seem straightforward to cross-compile `1.1.114`+ via MinGW, the added dependency on the D3D SDK makes things tricky.

I'll see upstream if it could be made optional for the time being.

The update instructions I put in `README.md` are thus not exact right now, since that's the workflow for `1.1.114` and later, but it should become relevant soon enough.

I noticed that we ship two `.asm` files which we don't use, and we compile `asm_offset.c` which seems only needed to generate `gen_defines.asm`. For now we could likely remove those three files, but ideally we should figure out how to use them in our own buildsystem.

**Comment by akien-mga** [comment]
I opened https://github.com/KhronosGroup/Vulkan-Loader/issues/215 about the WDK issue. For now it seems we're stuck on `1.1.113`...

**Comment by bruvzg** [comment]
>This makes things tricky for cross-compilation from Linux as this header is not part of mingw but is provided by WINE. In my distro it's at /usr/include/wine/windows/ddk/d3dkmthk.h which is not in the default include paths, so I'll have to hardcode this extra include path.

I can't find this header on Windows with Visual Studio 2019 installed either, probably it's from Windows Driver Kit.

**Comment by volcanoauthors** [comment]
I'm seeing the same issue with [ndsol/VolcanoSamples](https://github.com/ndsol/VolcanoSamples); renaming the symbols seems to be the legal route Microsoft suggests. What's the minimal rename that would pass their test? (Note, I'm only looking at the symbols actually used by loader/loader.c in vulkan-loader)

s/D3D/E2E/

Is that sufficient? Arguments to GetProcAddress are excluded, as per the Microsoft license language.

**Comment by volcanoauthors** [comment]
(The idea is to then copy the relevant header contents from the Microsoft distribution to somewhere else and do an identical symbol renaming process.)

**Comment by akien-mga** [comment]
> (The idea is to then copy the relevant header contents from the Microsoft distribution to somewhere else and do an identical symbol renaming process.)

See https://github.com/KhronosGroup/Vulkan-Loader/issues/215#issuecomment-519842580, as I understand it the documentation doesn't give clear permission to copy the header code from the WDK. It says more that you're allowed to reimplement it, and then it advises to use different symbol names to avoid name conflicts; in which case I'd add a prefix or suffix instead of doing a change such as `s/D3D/E2E/`.

**Comment by volcanoauthors** [comment]
It turns out it's really straightforward to copy/paste only the needed lines from docs.microsoft.com. Here's what we ended up with:

[winpatch.py to keep our windows build to a minimum](https://github.com/ndsol/volcano/blob/master/src/gn/winpatch.py#L95)

[raw patch from the posix build which may be more useful to you](https://github.com/ndsol/volcano/blob/master/src/gn/build-posix.sh#L381)

**Comment by akien-mga** [comment]
Nice, thanks for sharing it! I'll give it a try asap.

**Pull request merged**